### DISCUSSION TITLE
Possibility to correlate ast nodes to line and column number in visitor

### DISCUSSION THREAD

**Discussion by ghost** [swc-project/swc#1476] [Q&A] [answered]
Hello,

Thank you for the amazing work on swc, wanted to ask if it's possible to get line column number corresponding to nodes when travering the ast generated by swc.

**Comment by ghost** [comment]
well, i found the methods to convert Span to location in the rust code, these are not exposed to the web assembly version currently i guess.

**Comment by kdy1** [comment, marked as answer]
Currently, it's true. There's no api for it.

I'm considering something like `babelify` which will return babel ast to javascript.
Related to: https://github.com/swc-project/swc/pull/1465

**Reply by ghost** [reply]
@kdy1  would it be complex to expose the Span <-> source code location mapping API on the web assembly version? 

Something like `babelify` would be cool as well, as I guess then all the babel transforms out there can work with SWC as well, but I guess there will be performance penality of translation from SWC AST to babel AST and back.

**Reply by kdy1** [reply]
> would it be complex to expose the Span <-> source code location mapping API on the web assembly version?

No. It would be very straightforward.
I was worried about two api (babelify / resolveSpan or something simillar) to do same thing.

**Reply by ghost** [reply]
@kdy1  ah, understood. It makes sense better to have one proper API than a bunch of them. 

If it is very very straightforward could you maybe write a sample snippet or so how I could expose let's say resolveSpan to resolve Spans on AST returned from parseSync. I could build the wasm module myself with the changes and use it until the official solution is in. I'm very new to rust and swc and is still trying to figure out the codebase.

**Reply by ghost** [reply]
@kdy1 actually nevermind my last request, i have already figured it out and it works perfectly. Thanks again for the awesome work on SWC!!.

**Reply by sam-goodwin** [reply]
@jiby-gh Mind sharing your solution? I also need to map Spans to line/column.

**Reply by gb-jos** [reply]
@sam-goodwin its was a while ago, but I believe what I used was `node.span.lo.0`

**Reply by sam-goodwin** [reply]
Thanks for the quick reply, I actually figured it out. The entrypoint to the plugin is passed a `TransformPluginProgramMetadata` which contains a `PluginSourceMapProxy` which can be used to map a `BytePos` such as `span.lo()` to a `line_col`

```ts
#[plugin_transform]
pub fn my_plugin(mut program: Program, meta: TransformPluginProgramMetadata) -> Program {
  // you need this
  let source_map = meta.source_map;

  // then, given a span
  let span = ??
 
  // map it to a line_col
  let line_col = source_map.lookup_char_pos(span.lo());
  line_col.line; // 1-based line
  line_col.col_display; // 0-based column

  program
}
```
